cpp/src/arrow/csv/options.cc

```diff
@@ -17,11 +17,32 @@
 
 #include "arrow/csv/options.h"
 
+#include <iomanip>
+
 namespace arrow {
 namespace csv {
 
 ParseOptions ParseOptions::Defaults() { return ParseOptions(); }
 
+Status ParseOptions::Validate() const {
+  if (ARROW_PREDICT_FALSE((delimiter < ' ' && delimiter != '\t') || delimiter > '~')) {
+    return Status::Invalid(
+        "ParseOptions: delimiter must be a printable ascii char or '\\t': 0x",
+        std::setfill('0'), std::setw(2), std::hex, static_cast<uint16_t>(delimiter));
```

Comment:
Hmm, I don't think I understand this check. Is there something in the CSV parser that currently prevents using other delimiters?

Reply:
A lot of the characters before ` ` are terminal control characters and the only ascii character after `~` is `del`. While these characters could be used in theory for a text file they would be very strange couldn't be displayed on a terminal. These are the characters this is excluding https://flaviocopes.com/non-printable-ascii-characters/ . Also, I don't think the user could specify values like `\n` or `\r` since that would break the parser since they have to be used for end of line. Currently the python layer requires all character values to be between 1 - 127 inclusive.
I kept tab because tab separated values is a format which is used.

Reply:
I don't think it's our duty to guard against dubious values. Perhaps some weird formats use a form-feed character as a delimiter, who knows?

The only think that may be reasonable may be to forbid `\n` and `\r`. Otherwise we should just let the user choose whatever they like.